### DISCUSSION TITLE
Local calendar, calendar event mutation APIs, and recurring events

### DISCUSSION THREAD

**Discussion by allenporter** [home-assistant/architecture#797] [Entity Models] [answered]
# Proposal

Add new websocket APIs and `CalendarEntity` async methods to support `create`, `update`, and `cancel` of calendar events. All APIs will  be modeled after rfc5545 (and rfc5546 where possible), and will also support recurring events.

A new integration `local_calendar` implements fully functioning rfc5545 compatible calendar components. The proposed APIs will also work for `google` (and probably `caldav`).

Not in scope: This proposal covers a minimal set of event component properties needed, but obviously any other rfc5545 properties (e.g. `geo`, `attendees`, etc) can be added to `CalendarEvent` and are not discussed here.

# APIs & Identifiers

The update and cancel calls support operating on objects like:
- A single event
- All events in a recurring event series
- On a single event in a recurring event series
- On all events at and after a single event in a recurring event series

A calendar event is identified by a `uid` which is a globally unique identifier. A `uid` identifies a recurring event and all events in the series. A recurring event is expanded to multiple events and each instance has a `recurrence_id` unique to that date/time that identifies the individual instance.

A `update` or `cancel` call can specify a `uid` to refer to an event or all events in the series. Additionally a `recurrence_id` present will update or cancel a single event. ([rfc5546](https://www.rfc-editor.org/rfc/rfc5546.html#section-3.7.1))

An optional `recurrence_range` field can be set to `THISANDFUTURE` to also update or cancel all future events ([spec](https://www.rfc-editor.org/rfc/rfc5545#section-3.2.13)) based on the `recurrence_id`.

# Recurring events

Recurring events are complex. Our principals are:
- Home assistant core itself will not ever evaluate recurring events, other than acting as a pass through between clients (frontend) and the integration.
- Only [RECUR](https://www.rfc-editor.org/rfc/rfc5545#section-3.3.10) values that make sense for home assistant are supported (e.g. `FREQ` of `YEARLY`, `MONTHLY`, `WEEKLY`, `DAILY`)

A calendar integration is responsible for its own expansion of recurring events, and must return expanded single events from all APIs. This is already the status quo, but not explicitly written down anywhere.

`CalendarEntity will be updated with a new property `rrule` string which is an rfc5545 RRULE such as `FREQ=WEEKLY;UNTIL=20221231`. Home assistant core will not interpret the rrule, which will be passed straight through the APIs so that clients like the frontend can show a picker to modify the rule.

The rrule spec in rfc5545 will only be partially supported, just for what makes sense for our use cases.. Only FREQ values of `YEARLY`, `MONTHLY`, `WEEKLY`, `DAILY` and ignore all the others since they are not practical. An unsupported RRULE will not be modify-able by the frontend, but really core won't care.

# Implementation Details

I've implemented  the above add / update / cancel semantics in a python rfc5545 parsing/encoding library, and a proof of concept of a local calendar working in a home assistant with a create event service.

I've verified that the above proposal seems like it should work with google calendar [recurring events](https://developers.google.com/calendar/api/guides/recurringevents?hl=en#creating_recurring_events), and haven't looked at caldav in detail yet.

# Alternatives Considered

An alternative could be not to do this, and say that mutations should happen outside of Home Assistant using some  integration specific client (e.g. Google Calendar webside, caldav client). Local calendar could rely entirely on `caldav` an improve community support of some popular caldav server add-on.

**Comment by Olen** [comment]
I have created a PR to add event status to the CalendarEvent class. 

https://github.com/home-assistant/core/pull/77917

It should be in line with the RFCs. 

This will allow the different calendar integrations to set event status (typically "tentative", "confirmed" or "cancelled") to individual events, and use the status automation triggers and conditions.

**Comment by allenporter** [comment]
I have a proof of concept working and thought it might be time to reflect on the proposal above at another level of detail.

The calendar entity will need a new `CalendarEntityFeature` to indicate it supports create, update, cancel.

We first need to update the `CalendarEvent` schema to include additional identifier fields as well as recurring event fields:
```
CALENDAR_EVENT_SCHEMA = vol.Schema(
    {
        vol.Optional(CONF_UID): cv.string,
        vol.Optional(CONF_RECURRENCE_ID): cv.string,
        vol.Optional(CONF_START): vol.Any(cv.datetime, cv.date),
        vol.Optional(CONF_END): vol.Any(cv.datetime, cv.date),
        vol.Optional(CONF_SUMMARY): cv.string,
        vol.Optional(CONF_DESCRIPTION): cv.string,
        vol.Optional(CONF_LOCATION): cv.string,
        vol.Optional(CONF_RRULE): cv.string,
    }
)
```

We also need some new websocket operations:
```
type: "calendar/event/create"
entity_id: calendar.example
event:
  # Calendar event schema above
```

Delete:
```
type: "calendar/event/cancel"
entity_id: calendar.example
uid: <event id>
recurrence_id: <recurrence_id>
recurrence_range: THISANDFUTURE
```

These will correspond to new `CalendarEntity` optional methods:
```
    async def async_create_event(self, **kwargs: Any) -> dict[str, Any]:
        """Add a new event to calendar."""

    async def async_cancel_event(
        self,
        uid: str,
        recurrence_id: str | None = None,
        recurrence_range: str | None = None,
    ) -> None:
        """Cancel an event on the calendar."""

```

(Omitting update for now, but it is a hybrid of create and delete)

Here is a demo with a poor frontend implementation so you can get an idea of the operations that might be supported:

https://user-images.githubusercontent.com/6026418/192944153-7b5069c5-4416-4f26-9d19-6dfc2deae4e7.mov


I think the breakdown for initial PRs in core could look something like this:
- Make Google Calendar existing service call generic as part of `calendar`
- Initial `local_calendar` implementation using service calls to create events
- Add new `calendar` create/cancel websocket APIs, targeting local calendar as first implementation
- Implement create/cancel websocket APIs for Google calendar

**Reply by Olen** [reply]
Looks very good.
But I think you should make a distinction between "cancel event" and "delete event".  To me, a cancelled event is still present in the calendar backend, but might or might not be visible in the frontend depending on the users preferences.
Whereas a deleted event is removed from the backend as well.

**Reply by allenporter** [reply]
Good point given our discussion about status. So consider everything here about deletion and I recommend not supporting cancelation (right now for sure, and likely not needed in the future either).

**Comment by balloob** [comment, marked as answer]
This is good to go 👍
